nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/pgp/PGPControllerService.java

```diff
@@ -0,0 +1,893 @@
+            throw new IOException("Sign operation invalid without Private Key");
+        }
+        final int KEY_ALGORITHM = 1; // wha
+        JcaPGPContentSignerBuilder builder = new JcaPGPContentSignerBuilder(KEY_ALGORITHM, options.getHashAlgorithm()).setProvider("BC");
+        PGPSignatureGenerator generator = new PGPSignatureGenerator(builder);
+        generator.init(PGPSignature.BINARY_DOCUMENT, options.getPrivateKey());
+        readAndUpdate(input, generator);
+        generator.generate().encode(signature);
+    }
+
+
+    /**
+     * Read from an {@link InputStream} to verify its signature.
+     *
+     * @param input clear or cipher data
+     * @param signature signature data
+     * @param options used to configure verify operation
+     * @return true if the signature matches
+     */
+    public boolean verify(InputStream input, InputStream signature, VerifyOptions options) throws IOException, PGPException {
+        JcaPGPObjectFactory signatureFactory = new JcaPGPObjectFactory(org.bouncycastle.openpgp.PGPUtil.getDecoderStream(signature));
+        PGPSignatureList signatures;
+
+        Object head = signatureFactory.nextObject();
+        if (head instanceof PGPCompressedData) {
+            PGPCompressedData compressed = (PGPCompressedData) head;
+            signatureFactory = new JcaPGPObjectFactory(compressed.getDataStream());
+            signatures = (PGPSignatureList) signatureFactory.nextObject();
+        } else if (head instanceof  PGPSignatureList) {
+            signatures = (PGPSignatureList) head;
+        } else {
+            throw new PGPException("Not a signature stream.");
+        }
+
+        PGPSignature innerSignature = signatures.get(0);
+        innerSignature.init(new JcaPGPContentVerifierBuilderProvider().setProvider("BC"), options.getPublicKey());
+        readAndUpdate(input, innerSignature);
+        return innerSignature.verify();
+    }
+
+
+    /**
+     * Generate an {@link EncryptOptions} instance for an encrypt operation.
+     *
+     * @param algorithm encryption algorithm identifier, see {@link org.bouncycastle.bcpg.SymmetricKeyAlgorithmTags}
+     * @param armor if true, encrypted data will be PGP-encoded ascii
+     * @return encrypt options
+     */
+    @Override
+    public EncryptOptions optionsForEncrypt(int algorithm, boolean armor) {
+        PGPEncryptedDataGenerator generator = new PGPEncryptedDataGenerator(new JcePGPDataEncryptorBuilder(algorithm)
+                .setWithIntegrityPacket(true)
+                .setProvider("BC"));
+        char[] passphrase = getPBEPassPhrase(getConfigurationContext());
+        if (passphrase != null && passphrase.length != 0) {
+            generator.addMethod(new JcePBEKeyEncryptionMethodGenerator(passphrase).setProvider("BC"));
+        } else {
+            generator.addMethod(new JcePublicKeyKeyEncryptionMethodGenerator(getPublicKey(getConfigurationContext())).setProvider("BC"));
+        }
+
+        return new EncryptOptions() {
```

Comment:
Is there a reason to make this an anonymous inner class rather than a concrete implementation of the interface for reusability?

Reply:
Yes:  there are only 2 cases each for EncryptPGP and DecryptPGP, and I think it's much more obvious to the reader and maintainer what is happening, where, and why, than having the code split into smaller classes.  And the code is small.